external-crates/move/crates/move-compiler/src/typing/macro_expand.rs

```diff
@@ -0,0 +1,955 @@
+        // `|i| i``
+        // In this case we have
+        // two different colored `i`s when applying the outer lambda, e.g.
+        // `let i#_#c = arg; v.push_back({ let i#_#d = i#_#c; i#_#d })`
+        // we need to make sure `i#_#c` and `i#_#d` remain separated
+        //
+        // This has similar feeling to lifting  De Bruijn indices, though it is not exactly the same
+        // (... I think)
+        pub fn remap_color(&mut self, color: Color) -> Color {
+            *self.remapping.entry(color).or_insert_with(|| {
+                let cur = self.next_color;
+                self.next_color += 1;
+                cur
+            })
+        }
+
+        pub fn recolor_use_funs(&self) -> bool {
+            self.recolor_use_funs
+        }
+
+        pub fn max_color(&self) -> Color {
+            if self.remapping.is_empty() {
+                // next color never used
+                self.next_color
+            } else {
+                // subtract one to skip the "next" color
+                let max = self.next_color - 1;
+                debug_assert!(self.remapping.values().all(|&c| c <= max));
+                max
+            }
+        }
+
+        pub fn return_label(&self) -> Option<BlockLabel> {
+            self.return_label
+        }
+
+        pub fn contains_var(&self, v: &Var) -> bool {
+            self.vars.contains(v)
+        }
+
+        pub fn contains_block_label(&self, label: &BlockLabel) -> bool {
+            self.block_labels.contains(label)
+        }
+    }
+}
+
+fn recolor_var_owned(ctx: &mut Recolor, mut v: Var) -> Var {
+    assert!(ctx.contains_var(&v));
+    recolor_var(ctx, &mut v);
+    v
+}
+
+fn recolor_var(ctx: &mut Recolor, v: &mut Var) {
+    // do not recolor if not in the ctx
+    // this is to handle captured variables in lambda bodies
+    if !ctx.contains_var(v) {
+        return;
+    }
+    v.value.color = ctx.remap_color(v.value.color);
+}
+
+fn recolor_block_label_owned(ctx: &mut Recolor, mut label: BlockLabel) -> BlockLabel {
+    assert!(ctx.contains_block_label(&label));
+    recolor_block_label(ctx, &mut label);
+    label
+}
+
+fn recolor_block_label(ctx: &mut Recolor, label: &mut BlockLabel) {
+    // do not recolor if not in the ctx
+    // this is to handle captured labels in lambda bodies
+    if !ctx.contains_block_label(label) {
+        return;
+    }
+    label.label.value.color = ctx.remap_color(label.label.value.color);
+}
+
+fn recolor_use_funs(ctx: &mut Recolor, use_funs: &mut UseFuns) {
+    recolor_use_funs_(ctx, &mut use_funs.color);
+}
+
+fn recolor_use_funs_(ctx: &mut Recolor, use_fun_color: &mut Color) {
+    if ctx.recolor_use_funs() {
+        assert_eq!(
+            *use_fun_color, 0,
+            "ICE only expected to recolor use funs in fresh macro bodies"
+        );
+        *use_fun_color = ctx.remap_color(*use_fun_color);
+    }
+}
+
+fn recolor_seq(ctx: &mut Recolor, (use_funs, seq): &mut N::Sequence) {
+    recolor_use_funs(ctx, use_funs);
```

Comment:
I wonder if it would be worthwhile to rename all of this to something like:

```suggestion
    recolor_scope(ctx, use_funs);
```

And ultimately replace the `UseFuns` form with ` ColorScope` form. It might better-inform what's going on across this + core + translation. Not a huge deal, just thinking out loud.

Reply:
It's hard because it also holds the use_funs. So its like a method scope plus a color. Given methods+color, thoughts on a name?

Reply:
We can bikeshed more later. I wonder if `method_color_scope` or something, but this isn't show-stopping.